docs/proposals/alternate-api-representations.md

```diff
@@ -0,0 +1,422 @@
+    which matches our current use of the protobuf, JSON, and YAML serializations
+  * We do not wish to alter the mechanism for authorization - a user with access
+    to a particular resource in a given namespace should be limited regardless of
+    the representation in use.
+  * Allowing "all namespaces" to be listed would require us to create "fake" resources
+    which would authorization
+* We wish to support retrieving object representations in multiple schemas - JSON for
+  simple clients and Protobuf for clients concerned with efficiency.
+* Most clients will wish to retrieve the more efficient / simpler object, but for
+  older servers will desire to fall back to the implict resource represented by
+  the endpoint.
+  * Over time, clients may need to request results in multiple API group versions
+    because of breaking changes (when we introduce v2, clients that know v2 will want
+    to ask for v2, then v1)
+* We wish to preserve the greatest possible query parameter space for sub resources
+  and special cases, which encourages us to avoid polluting the API with query
+  parameters that can be otherwise represented.
+* Because we expect not all extensions will implement protobuf, an efficient client
+  must continue to be able to "fall-back" to JSON, such as for third party
+  resources.
+* We do not wish to create fake content-types like `application/json+kubernetes+v1+api.k8s.io`
+  because the list of combinations is unbounded and our ability to encode specific values
+  (like slashes) into the value is limited.
+
+### Client negotiation of response representation
+
+When a client wishes to request an alternate representation of an object, it should form
+a valid `Accept` header containing one or more accepted representations, where each
+representation is represented by a media-type and [media-type parameters](https://tools.ietf.org/html/rfc6838#section-4.3).
+The server should omit representations that are unrecognized or in error - if no representations
+are left after omission the server should return a `406 Not Acceptable` HTTP response.
+
+The supported parameters are:
+
+| Name | Value | Default | Description |
+| ---- | ----- | ------- | ----------- |
+| g | The group name of the desired response | Current group | The group the response is expected in. |
+| v | The version of the desired response | Current version | The version the response is expected in. Note that this is separate from Group because `/` is not a valid character in Accept headers. |
+| as | Kind name | None | If specified, transform the resource into the following kind (including the group and version parameters). |
+| sv | The server group (`api.k8s.io`) version that should be applied to generic resources returned by this endpoint | Matching server version for the current group and version | If specified, the server should transform generic responses into this version of the server API group. |
```

Comment:
sv controls the schema I get back on errors (Status in api.k8s.io/v1).  If a generic client asks for a response in a particular group version it may still want errors in a known schema, and for back compatibility we can't return Status in the server group by default.